metrics/engine/ingester.go

```diff
@@ -92,6 +97,74 @@ func (oi *outputIngester) flushMetrics() {
 				oi.metricsEngine.markObserved(sm.Metric)
 				sm.Metric.Sink.Add(sample)
 			}
+
+			oi.cardinality.Add(sample.TimeSeries)
 		}
 	}
+
+	if hit, limit := oi.cardinality.LimitHit(); hit {
+		// TODO: suggest using the Metadata API as an alternative, once it's
+		// available (e.g. move high-cardinality tags as Metadata)
+		// https://github.com/grafana/k6/issues/2766
+
+		oi.logger.Warnf("The test has generated %d time series, which is higher than the suggested limit (%d), "+
+			"and could cause high memory usage. "+
+			"Consider not using high-cardinality values like IDs as tags "+
+			"or if you need them in the URL then use the URL-grouping feature. "+
+			"See https://k6.io/docs/using-k6/tags-and-groups for details.", oi.cardinality.Count(), limit)
+	}
+}
+
+type cardinalityControl struct {
+	seen            map[metrics.TimeSeries]struct{}
+	timeSeriesLimit int
+	addsCount       int
```

Comment:
I was initially confused by `addsCount`, but I guess you are only using it for optimization purposes? Is `len(map)` a an operation that actually requires some calculation? I thought it just reads an internal counter? So this additional counter here might be a slight overhead, not an optimization

In any case, if we decide to keep it, it should probably be a `uint64` or something like that, considering `int` might be just 32 bits